quantum/action_tapping.c

```diff
@@ -117,6 +117,75 @@ void action_tapping_process(keyrecord_t record) {
     }
 }
 
+/* Some conditionally defined helper macros to keep process_tapping more
+ * readable
+ */
+#    if (defined(AUTO_SHIFT_ENABLE) && defined(RETRO_SHIFT)) || defined(TAPPING_TERM_PER_KEY) || defined(PERMISSIVE_HOLD_PER_KEY) || defined(TAPPING_FORCE_HOLD_PER_KEY) || defined(HOLD_ON_OTHER_KEY_PRESS_PER_KEY)
+#        define DEFINE_tapping_keycode uint16_t tapping_keycode = get_record_keycode(&tapping_key, false)
```

Comment:
That define is rather awkward in my opinion. The original if-guarded use of `uint16_t tapping_keycode = ...` at the beginning of `process_tapping` is more readable and more natural.

Reply:
My strategy was to eliminate all explicit uses of tapping_keycode from process_tapping. Because the variable is only declared conditionally, all its uses also have to be guarded by compile-time conditions, which was exactly what I was trying to eliminate. It made sense to made all the compile-time conditions in one place, so I wrapped this conditional variable declaration in a macro as well.

Reply:
Another way to do this without having a giant `#if`:

```c
#    ifdef RETRO_TAPPING_PER_KEY
#        define NEED_tapping_keycode
...
#    endif

#    ifdef PERMISSIVE_HOLD_PER_KEY
#        define NEED_tapping_keycode
...
#    endif

...
```
and then inside `process_tapping()`:
```c
#ifdef NEED_tapping_keycode
    uint16_t tapping_keycode = get_record_keycode(&tapping_key, false);
#endif
```
This way the variable definition would stay inside the `process_tapping()` code, and the `#define NEED_tapping_keycode` lines would be near the macros that use that variable. (Having multiple `#define`s for the same macro is allowed if all of them define the macro to the same thing.)